soc/bouffalolab/bl60x/soc.c

```diff
@@ -0,0 +1,890 @@
+		system_set_root_clock(crystal == 32 ? 2 : 3);
+		sys_write32(MHZ(160), CORECLOCKREGISTER);
+		break;
+	case 4:
+		/* TODO: enable rom access 2T*/
+		system_set_root_clock_dividers(0, 1);
+		system_set_root_clock(crystal == 32 ? 2 : 3);
+		sys_write32(MHZ(192), CORECLOCKREGISTER);
+		break;
+	default:
+		break;
+	}
+
+	/* settle */
+	system_clock_delay_32M_ms(10);
+
+	/* we have PLL now */
+	system_set_PKA_clock(1);
+}
+
+static void uart_set_clock_enable(uint32_t enable)
+{
+	uint32_t tmp = 0;
+
+	if (enable) {
+		enable = 1;
+	}
+
+	tmp = sys_read32(GLB_BASE + GLB_CLK_CFG2_OFFSET);
+	tmp = (tmp & GLB_UART_CLK_EN_UMSK)
+	    | (enable << GLB_UART_CLK_EN_POS);
+	sys_write32(tmp, GLB_BASE + GLB_CLK_CFG2_OFFSET);
+}
+
+/* Clock:
+ * FCLK: 0
+ * 160 Mhz PLL: 1
+ * When using PLL root clock, we can use either setting, when using the 32Mhz
+ * Oscillator with a uninitialized PLL, only FCLK will be available.
+ */
+static void uart_set_clock(uint32_t enable, uint32_t clock, uint32_t divider)
+{
+	uint32_t tmp = 0;
+
+	if (divider > 0x7) {
+		divider = 0x7;
+	}
+
+	if (clock) {
+		clock = 1;
+	}
+	/* disable uart clock */
+	uart_set_clock_enable(0);
+
+	tmp = sys_read32(GLB_BASE + GLB_CLK_CFG2_OFFSET);
+	tmp = (tmp & GLB_UART_CLK_DIV_UMSK)
+	    | (divider << GLB_UART_CLK_DIV_POS);
+	sys_write32(tmp, GLB_BASE + GLB_CLK_CFG2_OFFSET);
+
+	tmp = sys_read32(HBN_BASE + HBN_GLB_OFFSET);
+	tmp = (tmp & HBN_UART_CLK_SEL_UMSK)
+	    | (clock << HBN_UART_CLK_SEL_POS);
+	sys_write32(tmp, HBN_BASE + HBN_GLB_OFFSET);
+
+	uart_set_clock_enable(enable);
+}
+
+
+/* TODO: should take crystal type, defaults to 40Mhz crystal for BL602 as seems
+ * the most common
+ */
+#define USE_CRYSTALL_32MHZ 0
+static void system_clock_init(void)
+{
+#if USE_CRYSTALL_32MHZ
+	system_init_root_clock(0, 32);
+	system_set_root_clock_dividers(0, 0);
+	system_clock_trim_32M();
+#else
+	system_init_root_clock(5, 4);
+	system_set_root_clock_dividers(0, 2);
+#endif
+	system_set_machine_timer_clock(1, 0, mtimer_get_clk_src_div());
+}
+
+static void peripheral_clock_init(void)
+{
+	uint32_t regval = sys_read32(0x40000024);
+	/* enable UART0 clock routing */
+	regval |= (1 << 16);
+	/* enable I2C0 clock routing */
+	regval |= (1 << 19);
+	sys_write32(regval, 0x40000024);
+	uart_set_clock(1, 1, 0);
+}
+
+#ifdef CONFIG_RISCV_GP
+ulong_t __soc_get_gp_initial_value(void)
+{
+	extern uint32_t __global_pointer$;
+	return (ulong_t)&__global_pointer$;
+}
+#endif
+
+/**
+ * @brief Perform basic hardware initialization at boot.
+ *
+ * This needs to be run from the very beginning.
+ * So the init priority has to be 0 (zero).
+ *
+ * @return 0
+ */
+static int bl_riscv_init(void)
+{
+	uint32_t key;
+	uint32_t *p;
+	uint32_t i = 0;
+	uint32_t tmp = 0;
+
+	key = irq_lock();
+
+	/* disable hardware_pullup_pull_down (reg_en_hw_pu_pd = 0) */
+	tmp = sys_read32(HBN_BASE + HBN_IRQ_MODE_OFFSET);
+	/* "BL_CLR_REG_BIT" */
+	tmp = tmp & HBN_REG_EN_HW_PU_PD_UMSK;
+	sys_write32(tmp, HBN_BASE + HBN_IRQ_MODE_OFFSET);
```

Comment:
CC: @VynDragon ?

Reply:
Hello, because those are not for pinctrl or GPIO, but likely for the flash (undocumented), being located in HBN configuration. This may even be removed entirely later (it is present in version 1.4.5 of the SDK, but not in the later version as this is probably the hardware default).

Reply:
Is this required to use printk on early stages of boot? I mean, before kernel drivers be available.

Reply:
So this changes the default pin assignment before the pinctrl driver started.
Glad to know, thank you!

Reply:
> Is this required to use printk on early stages of boot? I mean, before kernel drivers be available.
> So this changes the default pin assignment before the pinctrl driver started.

No, pinctrl does not control this, this is a HBN area and uses a completely different pin configuration for sleep wake (i think that's the most likely here, this is not a area i've explored much yet), which is wholly unrelated to normal pin operation (so does not affect printk unless there is some undocumented interaction), only affects some pins, and is used in the context of sleep or partial shut down, it is not documented on bl60x and bl70x, but has some documentation on bl61x.
The pinctrl driver is the only thing that affects pin assignement in normal running mode after the bootloader until GPIO driver is used, at the moment (unless you consider brown out detection/recovery = BOR to be related).